File: Task/Y-combinator/Ruby/y-combinator-2.rb
y = ->(f) {->(g) {g.(g)}.(->(g) { f.(->(*args) {g.(g).(*args)})})}

fac = ->(f) { ->(n) { n < 2 ? 1 : n * f.(n-1) } }

p 10.times.map {|i| y.(fac).(i)}

fib = ->(f) { ->(n) { n < 2 ? n : f.(n-2) + f.(n-1) } }

p 10.times.map {|i| y.(fib).(i)}

File: Task/Y-combinator/Scala/y-combinator-2.scala
val fac: Int => Int = Y[Int, Int](f => i => if (i <= 0) 1 else f(i - 1) * i)
fac(6)  //> res0: Int = 720

val fib: Int => Int = Y[Int, Int](f => i => if (i < 2) i else f(i - 1) + f(i - 2))
fib(6)  //> res1: Int = 8
